File: Cpp-DSA-Ejemplos/data_structures/linked_list_demo.cpp
#include <iostream>
#include <string>

// Definición de un Nodo
struct Node {
    std::string data;
    Node* next; // Puntero al siguiente nodo

    // Constructor
    Node(std::string val) : data(val), next(nullptr) {}
};

// Definición de la Lista Enlazada
class LinkedList {
private:
    Node* head; // Puntero al primer nodo de la lista

public:
    // Constructor
    LinkedList() : head(nullptr) {}

    // Destructor para liberar la memoria
    ~LinkedList() {
        Node* current = head;
        while (current != nullptr) {
            Node* next_node = current->next;
            delete current;
            current = next_node;
        }
        head = nullptr;
    }

    // Comprueba si la lista está vacía
    bool isEmpty() const {
        return head == nullptr;
    }

    // Añade un nodo al final de la lista
    void append(std::string data) {
        Node* new_node = new Node(data);
        if (isEmpty()) {
            head = new_node;
            return;
        }

        Node* current = head;
        while (current->next != nullptr) {
            current = current->next;
        }
        current->next = new_node;
    }

    // Añade un nodo al principio de la lista
    void prepend(std::string data) {
        Node* new_node = new Node(data);
        new_node->next = head;
        head = new_node;
    }

    // Elimina la primera ocurrencia de un nodo con el dato especificado
    void deleteNode(std::string data) {
        if (isEmpty()) {
            return;
        }

        // Si el nodo a eliminar es la cabeza
        if (head->data == data) {
            Node* temp = head;
            head = head->next;
            delete temp; // Liberar memoria del nodo eliminado
            return;
        }

        Node* current = head;
        // Recorre hasta encontrar el nodo anterior al que se quiere eliminar
        while (current->next != nullptr && current->next->data != data) {
            current = current->next;
        }

        // Si se encontró el nodo a eliminar
        if (current->next != nullptr) {
            Node* temp = current->next;
            current->next = temp->next;
            delete temp; // Liberar memoria del nodo eliminado
        } else {
            std::cout << "Dato '" << data << "' no encontrado en la lista." << std::endl;
        }
    }

    // Muestra los elementos de la lista
    void display() const {
        if (isEmpty()) {
            std::cout << "Lista vacía." << std::endl;
            return;
        }
        Node* current = head;
        while (current != nullptr) {
            std::cout << current->data << " -> ";
            current = current->next;
        }
        std::cout << "nullptr" << std::endl;
    }
};

// Función de demostración para la Lista Enlazada
void linked_list_demo() {
    LinkedList lista;
    std::cout << "¿Lista vacía? " << (lista.isEmpty() ? "Sí" : "No") << std::endl;

    std::cout << "\n--- Añadiendo elementos (append) ---" << std::endl;
    lista.append("A");
    lista.append("B");
    lista.append("C");
    lista.display();

    std::cout << "\n--- Añadiendo al inicio (prepend) ---" << std::endl;
    lista.prepend("Inicio");
    lista.display();
    std::cout << "¿Lista vacía? " << (lista.isEmpty() ? "Sí" : "No") << std::endl;

    std::cout << "\n--- Eliminando elementos ---" << std::endl;
    lista.deleteNode("B");
    std::cout << "Después de eliminar 'B':" << std::endl;
    lista.display();

    lista.deleteNode("Inicio");
    std::cout << "Después de eliminar 'Inicio':" << std::endl;
    lista.display();

    lista.deleteNode("Z"); // Intentar eliminar un elemento que no existe
    lista.display();
}

int main() {
    linked_list_demo();
    return 0;
}

File: Java-DSA-Ejemplos/data_structures/LinkedListDemo.java
package data_structures;

public class LinkedListDemo {
    Node head;

    public LinkedListDemo() {
        this.head = null;
    }

    public boolean isEmpty() {
        return head == null;
    }

    public void append(String data) {
        Node newNode = new Node(data);
        if (isEmpty()) {
            head = newNode;
            return;
        }

        Node lastNode = head;
        while (lastNode.next != null) {
            lastNode = lastNode.next;
        }
        lastNode.next = newNode;
    }

    public void prepend(String data) {
        Node newNode = new Node(data);
        newNode.next = head;
        head = newNode;
    }

    public void delete(String data) {
        if (isEmpty()) {
            return;
        }

        // Si el nodo a eliminar es la cabeza
        if (head.data.equals(data)) {
            head = head.next;
            return;
        }

        Node currentNode = head;
        while (currentNode.next != null && !currentNode.next.data.equals(data)) {
            currentNode = currentNode.next;
        }

        // Si se encontró el nodo a eliminar
        if (currentNode.next != null) {
            currentNode.next = currentNode.next.next;
        } else {
            System.out.println("Dato '" + data + "' no encontrado en la lista.");
        }
    }

    public void display() {
        if (isEmpty()) {
            System.out.println("Lista vacía.");
            return;
        }
        Node currentNode = head;
        StringBuilder sb = new StringBuilder();
        while (currentNode != null) {
            sb.append(currentNode.data).append(" -> ");
            currentNode = currentNode.next;
        }
        sb.append("None");
        System.out.println(sb.toString());
    }

    public static void main(String[] args) {
        LinkedListDemo lista = new LinkedListDemo();
        System.out.println("¿Lista vacía? " + lista.isEmpty());

        System.out.println("\n--- Añadiendo elementos (append) ---");
        lista.append("A");
        lista.append("B");
        lista.append("C");
        lista.display();

        System.out.println("\n--- Añadiendo al inicio (prepend) ---");
        lista.prepend("Inicio");
        lista.display();
        System.out.println("¿Lista vacía? " + lista.isEmpty());

        System.out.println("\n--- Eliminando elementos ---");
        lista.delete("B");
        System.out.println("Después de eliminar 'B':");
        lista.display();

        lista.delete("Inicio");
        System.out.println("Después de eliminar 'Inicio':");
        lista.display();

        lista.delete("Z"); // Intentar eliminar un elemento que no existe
        lista.display();
    }
}
